datafusion/physical-expr/src/scalar_function.rs

```diff
@@ -147,8 +149,19 @@ impl PhysicalExpr for ScalarFunctionExpr {
 
         if let ColumnarValue::Array(array) = &output {
             if array.len() != batch.num_rows() {
-                return internal_err!("UDF returned a different number of rows than expected. Expected: {}, Got: {}",
-                        batch.num_rows(), array.len());
+                // If the function is not volatile and all arguments are scalars,
+                // we can assume that returning a one-element array is equivalent to returning a scalar.
+                let preserve_scalar = array.len() == 1
+                    && self.fun.signature().volatility != Volatility::Volatile
```

Comment:
Why can we not do this for functions that are `Volatile`?

Reply:
A volatile function is supposed to return a different result on every invocation so I don't think it's safe to do that. The example we were discussing in the earlier PR is a hypothetical `rand(upper_bound: Int64) -> Int64` function.

Reply:
> I guess a bit confusing since that function is currently not implementable like you said in that thread.

I think so. It's currently not implementable. 

I think under the current UDF invoke model, we should preserve scalar even if it is volatile. For example, we have a UDF `fn random_string(len: int) -> String`, which returns a random string of length `len`. This function is `volatile`, but it should return a scalar if the len parameter is a scalar.

Reply:
Sure, I can make that change - was just leaning more on the safe side.